Source/levels/drlg_l1.cpp

```diff
@@ -1258,14 +1258,14 @@ void PlaceMiniSetRandom(const Miniset &miniset, int rndper)
 
 Point SelectChamber()
 {
-	int chamber;
-	if (!HasChamber1)
+	int chamber = 2;
+	if (!HasChamber1 && HasChamber2 && HasChamber3)
 		chamber = PickRandomlyAmong({ 2, 3 });
-	else if (!HasChamber2)
+	else if (HasChamber1 && !HasChamber2 && HasChamber3)
 		chamber = PickRandomlyAmong({ 3, 1 });
-	else if (!HasChamber3)
+	else if (HasChamber1 && HasChamber2 && !HasChamber3)
 		chamber = PickRandomlyAmong({ 2, 1 });
-	else
+	else if (HasChamber1 && HasChamber2 && HasChamber3)
 		chamber = GenerateRnd(3) + 1;

```

Comment:
How about reordering the tests so they're positive checks? That way we can explain the exceptional single-chamber case.

```cpp
int chamber;
if (HasChamber1 && HasChamber2 && HasChamber3) {
    chamber = GenerateRnd(3) + 1;
} else if (HasChamber1 && HasChamber2) {
    chamber = PickRandomlyAmong({ 2, 1 });
} else if (HasChamber1 && HasChamber3) {
    chamber = PickRandomlyAmong({ 3, 1 });
} else if (HasChamber2 && HasChamber3) {
    // Reversed list order compared to the previous 2 cases in order to match vanilla logic
    chamber = PickRandomlyAmong({ 2, 3 });
} else {
    // The dungeon generation logic ensures that chamber 2 is available if
    // either (or both of) 1 or 3 aren't, so if we ever end up with a single
    // chamber layout it's always chamber 2.
    chamber = 2;
}